### PR TITLE
feature: implement a gas price oracle based on indexed blocks

### PR DESCRIPTION
**Pull request by ptrus**

### CODE REVIEW

File: db/migrations/20220109122505_logs.up.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ type Log struct {
 	BlockHash string
 	TxHash    string `bun:",pk"`
 	TxIndex   uint
-	Index     uint `bun:",pk,allowzero"`
```
Comment (ptrus) [inline db/migrations/20220109122505_logs.up.go:47]: This is a no-op change, as `allowzero` tag was removed in one of the past versions, so it is ignored anyway (and not needed)

File: gas/backend.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	service.BackgroundService
+
+	// GasPrice returns the currently recommended minimum gas price.
+	GasPrice() *hexutil.Big
+}
+
+const (
+	// windowSize is the number of recent blocks to use for calculating min gas price.
+	// NOTE: code assumes that this is relatively small.
+	windowSize = 3
```
Comment (kostko) [inline gas/backend.go]: This seems quite small, should we increase it to something like 12 blocks or so?

File: gas/backend.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+var (
+	// minPriceEps is a constant added to the cheapest transaction executed in last windowSize blocks.
+	minPriceEps = *quantity.NewFromUint64(1_000_000_000) // 1 "gwei".
+
+	// defaultGasPrice is the default gas price reported if no better estimate can be returned.
+	//
+	// This is only returned when all calls to oasis-node for min gas price failed and there were no
+	// full blocks in last `windowSize` blocks.
+	defaultGasPrice = *quantity.NewFromUint64(10_000_000_000) // 10 "gwei".
```
Comment (kostko) [inline gas/backend.go]: Should this be updated to 100 gwei?

File: gas/backend.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	}
+	g.trackPrice(lastTxPrice)
+}
+
+func (g *gasPriceOracle) trackPrice(price *quantity.Quantity) {
+	// One item gets added to the prices array. Bump the current index
+	// for next iteration.
+	defer func() {
+		g.blockPricesCurrentIdx++
+		if g.blockPricesCurrentIdx >= windowSize {
```
Comment (kostko) [inline gas/backend.go]: Could use `g.blockPricesCurrentIdx = (g.blockPricesCurrentIdx + 1) % windowSize`.

File: indexer/utils.go [hunk 5]
```diff
@@ -248,6 +258,11 @@ func (ib *indexBackend) StoreBlockData(ctx context.Context, oasisBlock *block.Bl
 			return err
 		}
 
+		// Remember the gas price as a fallback in case parsing the last transaction would fail for whetever reason.
```
Comment (kostko) [inline indexer/utils.go]: This is not only a fallback, right? As if there are only Ethereum transactions, the above branch for the last transaction would not trigger.

File: tests/rpc/utils.go [hunk 6]
```diff
@@ -56,14 +57,16 @@ type Response struct {
...
+	OasisBlockTimeout = 60 * time.Second
 	GasLimit          = uint64(1_000_000) // Minimum gas limit required to pass all tests (not using gas estimation).
 )
 
 var (
-	db       *psql.PostDB
-	GasPrice = big.NewInt(10_000_000_000) // Minimum gas price: https://github.com/oasisprotocol/emerald-paratime/blob/07179c1d20eddad3b3c3c9373bc0c853ad57fa16/src/lib.rs#L36
-	w3       *server.Web3Gateway
+	db             *psql.PostDB
+	GasPrice       = big.NewInt(10_000_000_000) // Minimum gas price: https://github.com/oasisprotocol/emerald-paratime/blob/07179c1d20eddad3b3c3c9373bc0c853ad57fa16/src/lib.rs#L36
```
Comment (kostko) [inline tests/rpc/utils.go]: Should this be updated?

File: gas/backend.go (hunk 2 above)
Comment (ptrus) [reply gas/backend.go]: Agreed, maybe also the `fullblockThreshold` should be lowered a bit? Maybe 0.8?

Comment (kostko) [reply gas/backend.go]: Yeah sounds good.

File: tests/rpc/utils.go (hunk 6 above)
Comment (ptrus) [reply tests/rpc/utils.go]: Updated, no harm in updating before bumping the runtime version (which will be done in a separate PR).

File: indexer/utils.go (hunk 5 above)
Comment (ptrus) [reply indexer/utils.go]: True. Reworded the comment a bit.

File: gas/backend.go (hunk 3 above)
Comment (ptrus) [reply gas/backend.go]: Updated

**Review by kostko — APPROVED** [review]

**Pull request merged**